include/swift/AST/Type.h

```diff
@@ -150,7 +150,9 @@ enum class SubstFlags {
   /// Map member types to their desugared witness type.
   DesugarMemberTypes = 0x02,
   /// Substitute types involving opaque type archetypes.
-  SubstituteOpaqueArchetypes = 0x04
+  SubstituteOpaqueArchetypes = 0x04,
+  /// For \c mapTypeIntoContext, which should only map generic parameter types.
+  SkipArchetypes = 0x08,
```

Comment:
mapTypeIntoContext() used to assert if the input already had archetypes in it, to catch invariant violations. Is this no longer the case?

Reply:
No, the use of this flag in `mapTypeIntoContext` is supposed to disregard archetypes in the *substitutions*; we still assert that there aren't archetypes in the input type. Consider the following:

```swift
// An associated type can be resolved through a class conformance.
class Class<A>: P {}
protocol P {
  associatedtype A
}
protocol Q: P {
  func method(_: A)
}
extension Q {
  func test(arg: Q & Class1Generic<Self>) {
    arg.method(self) // [Self.A = type(of: self)]
  }
}
```

Reply:
When you get the interface type for `Q & Class<Self>`, `Self` should be a GenericTypeParamType there, not an archetype. If it's an archetype, it's a mistake to call mapTypeIntoContext().

Reply:
Perhaps there was a misunderstanding: to compute the member reference type for `arg.method`, I'm mapping the interface type of `method` into context using the generic signature of the contextual base type `Q & Class<Self>` (computed via `getOpenedArchetypeSignature`). Using the interface type of `arg` would give us `(Self) -> ()`, where `Self` is a GenericTypeParamType, which is wrong. The core of the issue is that [this code](https://github.com/apple/swift/blob/4871212e15ae38227ff37fdc4ec7e71adc77f20e/lib/AST/GenericEnvironment.cpp#L160-L162) is not expecting the concrete type to be an archetype from elsewhere. The hack-around with the flag is of course far from principled, and I will gladly hear out any suggestions.

Reply:
Ah, is the 'archetype from elsewhere' an opened type? Should mapTypeIntoContext() (and mapTypeOutOfContext()) be changed instead to ignore opened archetypes? Perhaps TypeBase::hasArchetype() should even return false for them, like it does for opaque result archetypes. Neither opened nor opaque result archetypes are "real" in the sense that they can be mapped back and forth to interface types without losing information.

I'm just slightly allergic to adding new flags to subst(). AllowLoweredTypes is already a hack that I've wanted to remove for years :)

Reply:
It's just a regular archetype (the type of `self` inside `test`). The same can be done a generic parameter other than `Self`, i.e.
```swift
extension Q {
  func test<T>(arg: Q & Class<T>, t: T) {
    arg.method(t) // t is an archetype
  }
}
```
Is there by chance an existing way to check whether an archetype belongs to a given generic environment? We could rewrite the assert in terms of that. 

> Should mapTypeIntoContext() (and mapTypeOutOfContext()) be changed instead to ignore opened archetypes?

This crossed my mind as well the other day when I was pushing `TypeResolutionStage::Contextual` toward extinction. The assertion was constantly (and, apparently, unreasonably) getting hit by types from the sil parser.

Reply:
> It's just a regular archetype (the type of self inside test).

Ok, in that case I'm still confused. It sounds like something else is going wrong here.

> No, the use of this flag in mapTypeIntoContext is supposed to disregard archetypes in the substitutions; we still assert that there aren't archetypes in the input type.

What do you mean by substitutions here? mapTypeIntoContext() only takes a single type, not a substitution map.

Reply:
> What do you mean by substitutions here?

The `TypeSubstitutionFn` that `mapTypeIntoContext` passes to `Type::subst`. 

> Ok, in that case I'm still confused. It sounds like something else is going wrong here.

Here's what happens. The generic signature of `Q & Class<X>` binds `Self.A` to whatever `X` is. In our case with `arg` as the base, it is a primary archetype. When we map the interface type of `method` (`(Self.A) -> Void`) into the context of the protocol composition, `Self` is substituted for its corresponding archetype first, then `getMemberForBaseType` is called, followed by `ArchetypeType::getNestedType`, and we end up in `getOrCreateArchetypeFromInterfaceType`, which discovers that this nested type `A` is bound to a concrete type (which happens to be a *foreign* archetype). Finally, it maps the concrete type into context (in case it has generic parameter types in structural positions), which hits the `!hasArchetype()` assert you already mentioned. 

It seems to be the problem is that the assert is not discriminating between archetypes that are and are not native to *this* generic environment.

Reply:
So if I'm understanding correctly, we're building a generic signature of the form

`<Self where Self : Q, Self : Class<some archetype>>`

?

It would be nice to avoid working with the existential's generic signature if possible. Can you refactor the code to do this differently?

Reply:
> <Self where Self : Q, Self : Class<some archetype>>

Yes

> It would be nice to avoid working with the existential's generic signature if possible. Can you refactor the code to do this differently?

I wonder. We still need a generic environment to map the type of the member into, or a generic signature to call `getCanonicalTypeInContext`. The latter evades the problem at the cost of losing sugar.

Reply:
The type of the member is mapped into the generic context of the body of the function it is referenced from after applying the base type's substitutions, isn't it?

Reply:
Why would it? The member comes from a different, unrelated context. Generic arguments of a member are usually replaced with type variables and inferred by the CS (the exception are members of existential types, where we currently replace `Self` with the base object type [here](https://github.com/apple/swift/blob/25f7145f2d3aa0049e289c9a38c6d56061c293a0/lib/Sema/ConstraintSystem.cpp#L1866-L1884)).

Reply:
In your example, you have `arg.method(self)`. The type of `arg` is some type involving archetypes from the generic environment of `Q`. We take the interface type of method() and apply the context substitution map for the type of `arg` to get the type of the member reference.

Generally we don't ever care about the archetypes of the generic context of the member being referenced, only the archetypes of the scope we're currently in.

Reply:
I was using the contextual type of the member so that I can erase covariant archetypes to their bounds. The reason I chose to do this with archetypes vs. type parameters is because `closeExistential` needs the same logic applied to types that contain archetypes. I will see if I can delay all these transformations until `closeExistential`.

Reply:
Ah, no, we have perform covariant erasure in `getTypeOfMemberReference`, or else the constraint system will make the wrong assumptions about the resulting type of the reference. We need a generic signature to erase covariant dependent member types, so I don't think we can avoid working with it.

Why are you hesitant about using the generic signature?

Reply:
> Ah, no, we have to perform covariant erasure in getTypeOfMemberReference, or else the constraint system will make the wrong assumptions about the resulting type of the reference. We need a generic signature to erase covariant dependent member types, so I don't think we can avoid working with it.

That's fine, but you shouldn't mapIntoContext() using that generic signature. You can go directly from an interface type to an erased type involving existential types.

> Why are you hesitant about using the generic signature?

Using the signature itself is probably fine, but you shouldn't use the GenericEnvironment of a signature that contains "concrete archetypes", because Type::subst(), mapTypeIntoContext() and so on aren't set up to handle it.

Reply:
> You can go directly from an interface type to an erased type involving existential types.

OK. I guess this means we're down to `getCanonicalTypeInContext()`.

Reply:
I updated f3a1baeba66a4caeb9e70d086402e29150d12a89 to use generic signature queries to erase the type instead and got rid of that hack with the flag. However, doing so revealed that a call to `Type::subst` is still made through `Solution::simplifyType()` during expression rewriting, when the fully opened type is simplified. I countered this in `GenericEnvironment::getOrCreateArchetypeFromInterfaceType` by not mapping the concrete type into context unless it has a type parameter, (a type parameter and a "concrete archetype" cannot appear in the concrete type of an equivalence class simultaneously).